community/rfcs/24-08-21-GenAIExample-002-Edge_Craft_RAG.md

```diff
@@ -0,0 +1,118 @@
+Ownership(TCO) and pipeline optimization techniques are the two main reasons
+to block this process.
+
+#### Total Cost of Ownership
+
+The HW requirement of a typical edge use case is a single host with one of the
+following combinations:
+- Intel(R) Core(TM) Ultra Processor
+- Intel(R) Core(TM) Processor + Intel(R) Iris(R) Xe Graphics
+- Intel(R) Core(TM) Processor + Intel(R) Arc(TM) A-Series Graphics
+- Intel(R) Xeon(R) Processor + Intel(R) Arc(TM) A-Series Graphics
+
+The scenarios with these hardware options block the edge users from using large 
+parameter size LLMs on-prem as well as sophisticated RAG pipeline for their data.
+Thus, the RAG pipeline at edge needs to be highly curated for underlying hardwares
+and suitable models accordingly.
+
+#### RAG Pipeline Optimization Techniques
+
+Tuning RAG pipeline is a systematic problem. First, the quality depends on the
+result of each stage in the pipeline as well as the end-to-end outcome. Second,
+optimization could be a trade-off among the metrics. It is difficult to decide one
+answer is better than another if it is slightly not accurate in number but more
+relevant to the query. Third, the optimization techniques may not intuitively
+reflect to metrics improvements. E.g., recrusive retrieval may contribute to
+improving the recall and context relevancy, or may not.
+
+### Motivation
+
+Edge Craft RAG (EC-RAG) is a customizable, tunable and production-ready
+Retrieval-Augmented Generation system for edge solutions. It is designed to
+curate the RAG pipeline to meet hardware requirements at edge with garanteed
+quality and performance.
+
+From quality perspective, EC-RAG is tunable in the indexing, retrieving,
+reranking and generation stages for particular edge use cases. From performance
+perspective, the pipeline is consolidated in a single service to eliminate the
+overhead of inter-service communication on a single host. Meanwhile, the inferencing
+stages like embedding, reranking and generation are optimized for Intel(R) Iris(R)
+Xe Graphics and Intel(R) Arc(TM) A-Series Graphics.
+
+### Design Proposal
+
+EC-RAG is composed of the following components:
+- UI for doc loading and interactive chatbot.
+- Gateway
+- Mega-service with a single micro-services for the tunable* EC-RAG pipeline.
+- LLM serving microservice optimized for Intel(R) Iris(R) Xe Graphics and Intel(R) Arc(TM) A-Series
+Graphics
+- VectorDB microservice optimized for Intel(R) Iris(R) Xe Graphics and/or Intel(R) Arc(TM) A-Series
+- Docker compose file to launch the UI, Mega/Micro-services
+
+> [!NOTE]
+> *Advanced tuning EC-RAG will need a tool co-piloting with the pipeline which will be described in
+> a separate doc
+
+Below diagram illustrates the overall components of EC-RAG:
+![EC-RAG Diagram](Edge_Craft_RAG.png)
+
+The EC-RAG pipeline will expose 3 types of REST API endpoint:
+- **/v1/data** for indexing
+- **/v1/settings** for configuration
```

Comment:
This endpoint is used for configuring online parameters, right? Could you provide a list of the configurable parameters here?

Reply:
parameters (vector DB params (chunk_size) and llm params (temperature) can be customized when using /v1/dataprep and /v1/chatqna. So the /v1/settings maybe not needed.

Reply:
The parameters for settings are not finalized yet, but it roughly includes 2 objects: pipeline object and model object.
A pipeline object is composed of:
- data ingestion method and parameters
- indexing method and parameters
- Retrieval method and parameters
- Post processing method and parameters
- Model object link
A model object is composed of: 3 models and their devices to run
- llm_id
- llm_device
- embedding_id
- embedding_device
- rerank_id
- rerank_device
I'll update the details ASAP.